### ISSUE TITLE
Entity ID uniqueness concern

### ISSUE THREAD

**Issue by hspaay** [FIWARE/tutorials.NGSI-LD#2]
First, thank you for writing these tutorials, they are a great help in getting started with NGSI-LD.

I think the uniqueness of the full entity IDs is in question when following the tutorial.

Reading section "Creating Several Entities at Once" there is a paragraph that says: 
> In both cases we have encoded each entity id according to the NGSI-LD specification - the proposal is that each id is a URN 
> follows a standard format: urn:ngsi-ld:<entity-type>:<entity-id>. This will mean that every id in the system will be unique.

Also important: probably better to name this 'service-id' or 'device-id' as the term entity-id can easily be confused with the entity's id itself.

The concern is that this does not make an entity's full id unique. There seems to be an assumption that there is a single creator of entities that can guarantee that 'entitiy-id' is unique for an entity-type. This is an incorrect assumption. In an IoT environment there can be multiple agents that publish a temperature for example, each agent will use its own entity-id but without central coordination there is a chance of duplication. Another example, in my test setup I have two zwave controllers. These controllers generate device-id's starting numbering at 1. Using various temperature sensors on both controllers show there are two sensors with device ID 2 simply because that is how the numbering per controller works out.

What seems to be missing is to include the unique ID of the publisher/agent that is responsible the entity creation. 
The entity ID format would look like this: 
> urn:ngsi-ld:`<entity-type>:<agent-id>:<device-id>`[:...]

As long as agent-id's are unique the entity will be unique.
There are other ways to solve this problem but way leaves a nice breadcrumb of ownership in the entity's full id.

I've seen this issue in various documents including the ETSI specs so this concern is wider spread. The tutorials however seems like a good place to start.

I hope this helps somebody. 
(updated to fix the entity ID presentation)

**Comment by jason-fox** [comment]
As you would expect, the early beginner tutorials are somewhat simplified so as to introduce topics to the reader at an appropriate rate. The uniqueness statement carries over from the **NGSI-v2** tutorials where indeed it _was_ possible to have the same `id` and distinguish by `type`, but this was to be discouraged for a beginner, it made queries messy.

> . There seems to be an assumption that there is a single creator of entities that can guarantee that 'entity-id' is unique for an entity-type. This is an incorrect assumption

I believe your scenario is correctly covered in the federated case where the broker would still have the one entity, but collate the data from multiple devices in the real world, so the attributes would be separated by `datasetId`. I would argue  that the assumption _"`id` is unique"_ is true in most cases, but this is  _"an incorrect assumption"_ in edge cases such as the scenario described above. The description in the text should therefore just be considered as  simplification to help the new user - there are no advanced federation tutorials yet.

It would be nice if the architects of the **NGSI-LD** brokers could clarify their position on this:
@kzangeli @ScorpioBroker @martin-p-bauer @bobeal

**Comment by martin-p-bauer** [comment]
Hi Jason,

NGSI-LD, based on JSON-LD and the linked data principles, assumes that Entity IDs are unique and they have to be URIs.
More specifically, you can use URNs, but any URI, also an opaque URI without any further internal structure is valid.
Ensuring the uniqueness of the URI is the job of the creator of the entity.

Unlike in NGSIv2, the type is NOT part of the entity ID, i.e. cannot be used for distinguishing different entities!

The assumption of the NGSI-LD API is that all information with the same Entity Id belongs to the same entity.
Information about the same entity can be stored in different places. If requested in a distributed or federated setting,
the different information of the entity stored in different places is aggregated and returned as a single entity.

If an entity, e.g. a car,  has the same attribute in two different places, e.g. the speed property of the car coming from both
the speedometer and the GPS, a datasetId can be added to uniquely identify the property. So with different datasetId
there can be any number of instances with the same attribute name. However, there can be only  one default attribute instance
without a datasetId for any attribute. If, on aggregation, there are two different instances without datasetId,
the more recent one is taken if it can be determined based on a timestamp (preferably observedAt), otherwise a random
choice is made by the Broker which default instance is to be returned.

Best regards,

Martin

------------------------------------------
Dr. Martin Bauer
マーティン・バウアー
Senior Researcher
NEC Laboratories Europe GmbH
Kurfürsten-Anlage 36
D-69115 Heidelberg
Tel: +49/ (0)6221/4342-168
Fax: +49/ (0)6221/4342-155
E-Mail: Martin.Bauer@neclab.eu<mailto:Martin.Bauer@neclab.eu>
http://www.neclab.eu

*************************************************************
NEC Laboratories Europe GmbH | Registered Office: Kurfürsten-Anlage 36, 69115 Heidelberg, Germany | Registered at Amtsgericht Mannheim Germany, HRB728558 | Managing Director: Dr. Jürgen Quittek

**Comment by hspaay** [comment]
Thanks for the clarification @jason-fox @martin-p-bauer. I didn't understand why entity type needed to be part of the the full entity ID and this clears it up. It isn't required. I didn't catch on the importance of datasetId to support multiple attribute instances so that is helpful as well.

As for uniqueness, as a developer of IoT agent-A I have to make sure the entity that is created has a unique URI. Agent-A however only knows about the devices it is creating entities for. Agent-B can create entities of similar types and will be unaware of Agent-A IDs. Therefore there must a way to distinguish an entity from agent A and agent B. Two options come to mind, 1) use UUIDs, 2) include the ID of the agent in the entity ID URI. The first option is makes querying ID useless while the second requires that agent IDs are unique. I understand that the NGSI-LD spec doesn't prescribe how to make IDs unique so we are on our own here.

@jason-fox including this use-case might indeed complicate the tutorials. It is a lot of work to keep things simple. However, when reading the tutorial one of the first questions that came up is how to keep entity IDs unique. The uniqueness statement in the tutorial throws me off as my first reaction is that I am missing something important. May I suggest to reword it to say that it is important that entity IDs are unique.

In [my iotdomain project](https://github.com/iotdomain/iotdomain-standard/blob/master/iotdomain-standard.md) uniqueness was guaranteed by including the domain and the publisher in the entity ID. Doing the same here would result in IDs that have this format:
> urn:ngsi-ld:`<domain>`:`<agent-ID>`:`<device-ID>`

Where:
* The domain is the scope of the data collected, spatial or functional, eg city, club. It could be the entity type as in NGSI-v2, I guess as long as the agent IDs are unique within the domain. 
* The agent-ID is the ID of the instance of the IoT agent and unique within the domain
* The device-ID is unique for the agent.

This is probably too much for the tutorial, but I'm just basing this on my experience with developing the iotdomain project. Anyone starting will eventually run into the question of how to define IDs that scale. Btw my intent is to convert the iotdomain project to work with NGSI-LD and every single publisher instance would be its own Agent. This means a lot of agents talking to the context broker. Good times :)
